### PR TITLE
52 ballot status

### LINKED ISSUE #52: Ballot Status
We need an enpoint which returns the `status` of the requestes ballot
- [x] how many people voted
- [x] what was voted and how much
- [x] ~how many tokens there are for the ballot (possible votes, #53)~

### PR DESCRIPTION
**Pull request by marcschwaiger**
Closes #52

### CODE REVIEW

File: src/handler/BallotHandler.ts [hunk 1]
```diff
@@ -47,4 +59,52 @@ export class BallotHandler {
 
     await ballotService.updateBallot(ballotID, running, question, options);
   }
+
+  public async getAdminStatus(
```
Comment (94726) [inline src/handler/BallotHandler.ts]: Choose a better name like `getVoteResult`

File: src/handler/BallotHandler.ts [hunk 2]
```diff
@@ -47,4 +59,52 @@ export class BallotHandler {
...
+    ballotID: string
+  ): Promise<TotalVoteCount> {
+    if (!(await ballotService.checkIfBallotIDExists(ballotID)))
+      throw new NotFound("Ballot with given id does not exist");
+
+    const votes = await voteService.getVotes(ballotID);
+
+    return {
+      count: votes.length
+    } as TotalVoteCount;
```
Comment (94726) [inline src/handler/BallotHandler.ts]: Don' use type assertion here. It should generally be avoided in most cases.

Comment (94726) [reply src/handler/BallotHandler.ts]: Also abstract more of this function into a service.
Handlers should optimally combine functions of services to handle conditional workflows like checking whether something exists.

File: src/handler/BallotHandler.ts (hunk 1 above)
Comment (94726) [reply src/handler/BallotHandler.ts]: Sorry cannot select lines on mobile, but most of this functions should be in a service, see other comment.

**Review by 94726 — CHANGES_REQUESTED** [review]

File: src/handler/BallotHandler.ts (hunk 1 above)
Comment (marcschwaiger) [reply src/handler/BallotHandler.ts]: Renamed in [#54 Moved counting logic to voting service and refactored](https://github.com/LeoTuet/backend-we-want-to-sleep/pull/54/commits/d2c6e6ec6461eb9eec43029b972964306cbd4c12)

File: src/handler/BallotHandler.ts (hunk 2 above)
Comment (marcschwaiger) [reply src/handler/BallotHandler.ts]: Resolved in [#54 Moved counting logic to voting service and refactored](https://github.com/LeoTuet/backend-we-want-to-sleep/pull/54/commits/d2c6e6ec6461eb9eec43029b972964306cbd4c12)

File: src/handler/BallotHandler.ts [hunk 3]
```diff
@@ -47,4 +49,24 @@ export class BallotHandler {
...
+  }
+
+  public async getTotalVoteCount(
+    ballotID: string
+  ): Promise<TotalVoteCount> {
+    if (!(await ballotService.checkIfBallotIDExists(ballotID)))
+      throw new NotFound("Ballot with given id does not exist");
+
+    return await voteService.countVotes(ballotID);
+  }
```
Comment (94726) [inline src/handler/BallotHandler.ts]: Should be in `VoteHandler.ts`

File: src/services/VoteService.ts [hunk 4]
```diff
@@ -36,4 +47,48 @@ export class VoteService {
       .map((option) => option.identifier)
       .includes(vote);
   }
+
+  public async countVotes(
+    ballotID: string
+  ): Promise<TotalVoteCount> {
+    if (await BallotRepository.getBallot(ballotID) == null)
+      throw new NotFound("Ballot with given id does not exist");
```
Comment (94726) [inline src/services/VoteService.ts]: Not sure if a service should do such logic. I once made an exception to this though in `BallotService.getRunningBallot`. You may change that the if check only happens in `BallotHandler.getRunningBallot` as I said in a comment above.

File: src/services/BallotService.ts [hunk 5]
```diff
@@ -15,11 +15,11 @@ export class BallotService {
...
     const ballot = await BallotRepository.getRunningBallot();
 
-    if (ballot) {
-      return ballot;
-    } else {
+    if (!ballot) {
       throw new createHttpError.NotFound("There is no running ballot");
     }
+
+    return ballot;
```
Comment (94726) [inline src/services/BallotService.ts]: Should be in `BallotHandler` now that I think about it.

File: src/controller/BallotController.ts [hunk 6]
```diff
@@ -64,4 +64,16 @@ export default {
       res.status(204).send()
     }
   ),
+  getVoteResult: asyncHandler(
+    async (req: Request<{ ballotID: string }, {}, {}>, res) => {
+      const result = await ballotHandler.getVoteResult(req.params.ballotID);
+      res.status(200).send(result);
```
Comment (94726) [inline src/controller/BallotController.ts]: Idk why but I think we agreed on always returning an object with `body: ...`
```suggestion
      res.status(200).send({
        body: result
      });
```

File: src/controller/BallotController.ts [hunk 7]
```diff
@@ -64,4 +64,16 @@ export default {
...
+    }
+  ),
+  getTotalVoteCount: asyncHandler(
+      async (req: Request<{ ballotID: string }, {}, {}>, res) => {
+          const totalCount = await ballotHandler.getTotalVoteCount(req.params.ballotID);
+          res.status(200).send(totalCount);
```
Comment (94726) [inline src/controller/BallotController.ts]:
```suggestion
          res.status(200).send({
            body: totalCount
          });
```

File: src/services/VoteService.ts [hunk 8]
```diff
@@ -36,4 +47,48 @@ export class VoteService {
...
+    }
+
+    // count votes
+    for (const vote of votes) {
+      results[vote.vote]++;
+    }
+
+    // fill records into array list
+    const out: VoteResult[] = [];
+    for (const result of Object.getOwnPropertyNames(results)) {
```
Comment (94726) [inline src/services/VoteService.ts]:
```suggestion
    for (const result of Object.keys(results)) {
```

File: src/services/VoteService.ts [hunk 9]
```diff
@@ -36,4 +47,48 @@ export class VoteService {
...
+      const voteOption = voteOptions.find(vo => result == vo.identifier)
+      const vr: VoteResult = {
```
Comment (94726) [inline src/services/VoteService.ts]:
```suggestion
      const voteResult = {
```

**Review by 94726 — CHANGES_REQUESTED** [review]

File: src/services/BallotService.ts (hunk 5 above)
Comment (marcschwaiger) [reply src/services/BallotService.ts]: Resolved in [#52 Moved getRunningBallot into BallotHandler](https://github.com/LeoTuet/backend-we-want-to-sleep/pull/54/commits/77e9d2ef9b5f570fc14e3f0bcd156119861eb6b6) and [#52 Added getRunningBallot into BallotService](https://github.com/LeoTuet/backend-we-want-to-sleep/pull/54/commits/d8760654a5630f8cf0b9490c885f696156f8c87e)

File: src/services/VoteService.ts (hunk 4 above)
Comment (marcschwaiger) [reply src/services/VoteService.ts]: Resolved in [#52 Removed unnecessary check in VoteService#countVotes](https://github.com/LeoTuet/backend-we-want-to-sleep/pull/54/commits/81b45bc6955ffe5c4e2b16b2e10f245399d5ebc2)

File: src/controller/BallotController.ts (hunk 7 above)
Comment (marcschwaiger) [reply src/controller/BallotController.ts]: Resolved in [#52 Using object with res.send in BallotController](https://github.com/LeoTuet/backend-we-want-to-sleep/pull/54/commits/91c0f570c9b931ffdb31633babb6f062b19a1f1e)

File: src/controller/BallotController.ts (hunk 6 above)
Comment (marcschwaiger) [reply src/controller/BallotController.ts]: Resolved in [#52 Using object with res.send in BallotController](https://github.com/LeoTuet/backend-we-want-to-sleep/pull/54/commits/91c0f570c9b931ffdb31633babb6f062b19a1f1e)

File: src/handler/BallotHandler.ts (hunk 3 above)
Comment (marcschwaiger) [reply src/handler/BallotHandler.ts]: Resolved in [#52 Moved getVoteResult and getTotalVoteCount into VoteHandler](https://github.com/LeoTuet/backend-we-want-to-sleep/pull/54/commits/7acdb4f7b092c75a612a0f0b0a4e01243fb094ee)

File: src/services/VoteService.ts (hunk 9 above)
Comment (marcschwaiger) [reply src/services/VoteService.ts]: Resolved in [#52 Renamed vr into voteResult](https://github.com/LeoTuet/backend-we-want-to-sleep/pull/54/commits/83b8907df1deead9c87cf87a9094963334be85ae)

File: src/handler/BallotHandler.ts [hunk 10]
```diff
@@ -32,7 +32,15 @@ export class BallotHandler {
 
   public getBallots = ballotService.getBallots;
 
-  public getRunningBallot = ballotService.getRunningBallot;
+  public getRunningBallot = async () => {
+    const ballot = await ballotService.getRunningBallot();
+
+    if (!ballot) {
+      throw new createHttpError.NotFound("There is no running ballot");
```
Comment (LeoTuet) [inline src/handler/BallotHandler.ts]: You can directly use the `NotFound` class like in line 28

Comment (marcschwaiger) [reply src/handler/BallotHandler.ts]: Resolved in [#52 Directly using NotFound in BallotHandler](https://github.com/LeoTuet/backend-we-want-to-sleep/pull/54/commits/cbc9572279ea49100e3dfb4376532fea2a5da678)

**Review by LeoTuet — APPROVED** [review]

File: src/handler/BallotHandler.ts [hunk 11]
```diff
@@ -32,7 +32,15 @@ export class BallotHandler {
 
   public getBallots = ballotService.getBallots;
 
-  public getRunningBallot = ballotService.getRunningBallot;
+  public getRunningBallot = async () => {
```
Comment (94726) [inline src/handler/BallotHandler.ts]: For gods sake, please just use a normal function.

File: src/services/VoteService.ts [hunk 12]
```diff
@@ -36,4 +46,45 @@ export class VoteService {
...
+  ): Promise<TotalVoteCount> {
+    const votes = await this.getVotes(ballotID);
+    return {
+      count: votes.length
+    };
+  }
+
+  public async getVoteResult(
+    ballotID: string,
+    voteOptions: VotingOption[]
```
Comment (94726) [inline src/services/VoteService.ts]: Don't see a reason why `ballotID` wouldn't be enough as you can query the voteOptions using the id directly.

**Review by 94726 — CHANGES_REQUESTED** [review]

File: src/handler/BallotHandler.ts (hunk 11 above)
Comment (marcschwaiger) [reply src/handler/BallotHandler.ts]: For god sake fix ur own code
Resolved in  [#52 For gods sake, I am using a normal method now](https://github.com/LeoTuet/backend-we-want-to-sleep/pull/54/commits/ec17f7ac247fcc6782d574760d8876a902d0009a)

File: src/services/VoteService.ts (hunk 12 above)
Comment (marcschwaiger) [reply src/services/VoteService.ts]: ok

**Comment by marcschwaiger** [comment]
Please end my suffering.

**Review by 94726 — APPROVED** [review]
Good job. Thanks for your help!

**Pull request merged**